### PR TITLE
Vendoring libnetwork v0.7.0-dev.10

### LINKED ISSUE #20335: Containers name resolution timed out with embedded DNS
I have two docker containers started and linked via `docker-compose.yml` file:

~~~
version: '2'
services:
  srv1:
    image: test-net
    expose:
      - 8000
    command: python -m SimpleHTTPServer
    working_dir: /tmp

  srv2:
    image: test-net
    links:
      - srv1
    command: tail -f /var/log/dpkg.log

networks:
  default:
    driver: bridge
    driver_opts:
      'com.docker.network.enable_ipv6': 'true'
~~~

`test-net` is a simple Ubuntu image with some additional stuff preinstalled.

After I run `docker-compose up`, I can attach to **srv2** container (`testdockercompose_srv2_1` in my example) and it looks like everything is set properly:

~~~
# cat /etc/resolv.conf

search haze.yandex.net yandex.net yandex.ru
nameserver 127.0.0.11
nameserver <my:host:system:ipv6:dns::1>
options ndots:0

# ping -w 5 127.0.0.11

PING 127.0.0.11 (127.0.0.11) 56(84) bytes of data.
64 bytes from 127.0.0.11: icmp_seq=1 ttl=64 time=0.090 ms
64 bytes from 127.0.0.11: icmp_seq=2 ttl=64 time=0.086 ms

--- 127.0.0.11 ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 3999ms
rtt min/avg/max/mdev = 0.065/0.081/0.090/0.011 ms

# ping -w 2 srv1

PING srv1 (172.19.0.2) 56(84) bytes of data.
64 bytes from testdockercompose_srv1_1.testdockercompose_default (172.19.0.2): icmp_seq=1 ttl=64 time=0.132 ms
64 bytes from testdockercompose_srv1_1.testdockercompose_default (172.19.0.2): icmp_seq=2 ttl=64 time=0.114 ms

--- srv1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1000ms
rtt min/avg/max/mdev = 0.114/0.123/0.132/0.009 ms
~~~

But if I try to call name resolution for `srv1` directly via `host` of `nslookup` I get "connection timed out":

~~~
# host srv1

;; connection timed out; no servers could be reached

# nslookup srv1

;; Got recursion not available from 127.0.0.11, trying next server
;; connection timed out; no servers could be reached
~~~

At the same time name resolution works fine if I specify DNS server explicitly:

~~~
# nslookup srv1 127.0.0.11

Server:		127.0.0.11
Address:	127.0.0.11#53

Non-authoritative answer:
Name:	srv1
Address: 172.19.0.2
~~~

This leads to HTTP timeouts during calls to linked containers. A simple curl call to `http://srv1:8000` takes 15 secs, while call to `http://172.19.0.2:8000` is instant.

My host system is IPv6-only Linux with docker 1.10.1 installed.

~~~
› docker version

Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

› sudo docker info

Containers: 9
 Running: 2
 Paused: 0
 Stopped: 7
Images: 88
Server Version: 1.10.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 93
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: host bridge null
Kernel Version: 3.19.0-42-lowlatency
Operating System: Ubuntu 14.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.954 GiB
Name: var01f.vsfront.yandex.net
ID: YYOZ:TMTI:UE6N:6DUW:S2GP:KGNN:TN3G:CEQO:F2CA:ODQE:FRQH:Q356
Debug mode (server): true
 File Descriptors: 33
 Goroutines: 54
 System Time: 2016-02-15T23:48:08.593476359+03:00
 EventsListeners: 0
 Init SHA1: e1042dbb0bcf49bb9da188176d9a5063cdb92a01
 Init Path: /usr/lib/docker/dockerinit
 Docker Root Dir: /var/lib/docker
WARNING: No swap limit support
Labels:
 provider=generic
~~~

### LINKED ISSUE #20448: docker 1.10 embedded DNS server forwarding container name requests to external DNS when using curl
Description of problem:

I use docker 1.10 and the new service discovery using the embedded DNS server. According the section *Docker embedded DNS server* of the [docker networking documentation](https://docs.docker.com/engine/userguide/networking/dockernetworks/), I understand that only unresolved requests are forwarded to the external DNS server. 

Unfortunately, by looking at my DNS server logs, I find out that some dns requests (when using curl for example) to a container name are forwarded to the external dns server defined on the host machine.

For example, with a simple nginx container named "myweb", the command `curl -XGET http://myweb` from another container on the same network produce the following entry in my DNS server logs:

```
18-Feb-2016 16:48:07.324 queries: info: client MY_HOST_IP#47069 (myweb): query: myweb IN AAAA + (MY_DNS_IP)
```

Strange thing is that a query to "myweb" using dig or nslookup does not appear in the DNS logs.

I tried bridge and overlay networks, using docker-compose 1.6.0 or docker directly and have always the same behavior.

I hope my explanations are clear enough

`docker version`:
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

`docker info`:
Containers: 22
 Running: 3
 Paused: 0
 Stopped: 19
Images: 101
Server Version: 1.10.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 294
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: host bridge null
Kernel Version: 3.19.0-25-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.797 GiB
Name: XXXXXXXXXXX
ID: FF3I:AT6Q:XQTR:TO6I:ZUB5:7MD2:X4E7:OFQP:T2CE:27IE:HPXA:5IZI
WARNING: No swap limit support

`uname -a`:
Linux XXXXXXXXXXX 3.19.0-25-generic #26~14.04.1-Ubuntu SMP Fri Jul 24 21:16:20 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Steps to Reproduce:

1. Setup your host to use a DNS server that you can access the queries logs
2. docker network create mynet
3. docker run -d --net=mynet --name myweb nginx
4. docker run -it --rm --net=mynet debian bash

then, inside the container:

1. apt-get update
2. apt-get install dnsutils curl
3. dig myweb
```
; <<>> DiG 9.9.5-9+deb8u5-Debian <<>> myweb
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 50747
;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;myweb.				IN	A

;; ANSWER SECTION:
myweb.			1800	IN	A	172.19.0.4

;; Query time: 0 msec
;; SERVER: 127.0.0.11#53(127.0.0.11)
;; WHEN: Thu Feb 18 16:03:02 UTC 2016
;; MSG SIZE  rcvd: 44
```

4. curl -XGET http://myweb
```
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

When using dig, the request seems not forwarded but when using curl, the following entry appears in my dns log:

```
18-Feb-2016 16:48:07.324 queries: info: client MY_HOST_IP#47069 (myweb): query: myweb IN AAAA + (MY_DNS_IP)
```

### PR DESCRIPTION
**Pull request by aboch**
Fixes #20335 
Fixes #20448

Signed-off-by: Alessandro Boch aboch@docker.com

### CODE REVIEW

File: integration-cli/docker_cli_network_unix_test.go [hunk 1]
```diff
@@ -1128,10 +1128,10 @@ func (s *DockerNetworkSuite) TestDockerNetworkConnectDisconnectToStoppedContaine
 
 func (s *DockerNetworkSuite) TestDockerNetworkConnectPreferredIP(c *check.C) {
 	// create two networks
-	dockerCmd(c, "network", "create", "--subnet=172.28.0.0/16", "--subnet=2001:db8:1234::/64", "n0")
+	dockerCmd(c, "network", "create", "--ipv6", "--subnet=172.28.0.0/16", "--subnet=2001:db8:1234::/64", "n0")
```
Comment (cpuguy83) [inline integration-cli/docker_cli_network_unix_test.go]: What effect does this flag have on the test?

**Comment by tonistiigi** [comment]
LGTM

**Comment by tonistiigi** [comment]
@aboch needs rebase

File: integration-cli/docker_cli_network_unix_test.go (hunk 1 above)
Comment (aboch) [inline integration-cli/docker_cli_network_unix_test.go]: `--ipv6` was added to master for `docker network create` in line what we have for `docker daemon`, but libnetwork was not honoring it fully.
With this vendoring, we are filling the gap: If ipv6 stack is not explicitly enabled for the network, the user specified ipv6 subnet will be ignored.

**Comment by aboch** [comment]
@tonistiigi Rebased. Thanks.

File: hack/vendor.sh [hunk 2]
```diff
@@ -29,7 +29,11 @@ clone git github.com/RackSec/srslog 259aed10dfa74ea2961eddd1d9847619f6e98837
 clone git github.com/imdario/mergo 0.2.1
 
 #get libnetwork packages
+<<<<<<< Updated upstream
```
Comment (cpuguy83) [inline hack/vendor.sh]: bad rebase here

Comment (aboch) [inline hack/vendor.sh]: Thanks, an unstaged changes...

File: integration-cli/docker_cli_network_unix_test.go (hunk 1 above)
Comment (aboch) [inline integration-cli/docker_cli_network_unix_test.go]: This was the docker PR: https://github.com/docker/docker/pull/17513
and this was the libnetwork part: https://github.com/docker/libnetwork/pull/912

**Comment by cpuguy83** [comment]
LGTM

**Comment by calavera** [comment]
All :green_heart:

**Pull request merged**